### DISCUSSION TITLE
Remote Datasets

### DISCUSSION THREAD

**Discussion by grabango-jwatters** [open-mmlab/mmdetection#9010] [Q&A] [answered]
What is the appropriate method to introduce a remote dataset into the `mmdet.apis.train_detector`? For example, if I have a Coco dataset, is there a way to pipeline `LoadImageFromFile` say from a cloud blob storage such as GCS? Another example would be the methods in which datasets and dataloaders are created as generators in the [webdataset](https://github.com/webdataset/webdataset) repository.

**Comment by jbwang1997** [comment, marked as answer]
You can write a new remote file client and register it into [mmcv.FileClient](https://github.com/open-mmlab/mmcv/blob/master/mmcv/fileio/file_client.py). Both`Dataset` and `LoadImageFromFile` can set the file client to load files from different backends.

**Reply by grabango-jwatters** [reply]
Thank you, @jbwang1997. The quick and dirty way I was able to achieve this feature can be seen below. Is there anything objective incorrect with the following implementation? Also, my understanding of the module registration is that it simply requires an import of the module prior to the training execution. Is that correct, is there more documentation around pipeline module registration?

```
import gcsfs
import mmcv
import numpy as np
from mmdet.datasets import PIPELINES


@PIPELINES.register_module()
class LoadImageFromGCS:
    def __init__(self,
                 to_float32=False,
                 color_type='color',
                 channel_order='bgr',
                 file_client_args=None):

        if file_client_args is None:
            file_client_args = dict()

        self.to_float32 = to_float32
        self.color_type = color_type
        self.channel_order = channel_order
        self.file_client_args = file_client_args.copy()
        self.file_client = None

    def __call__(self, results):

        if self.file_client is None:
            self.file_client = gcsfs.GCSFileSystem(**self.file_client_args)

        filename = results['img_info']['filename']

        with self.file_client.open(filename) as f:
            img_bytes = f.read()

        img = mmcv.imfrombytes(
        img_bytes, flag=self.color_type, channel_order=self.channel_order)

        if self.to_float32:
            img = img.astype(np.float32)

        results['filename'] = filename
        results['ori_filename'] = results['img_info']['filename']
        results['img'] = img
        results['img_shape'] = img.shape
        results['ori_shape'] = img.shape
        results['img_fields'] = ['img']
        return results
```
